### DISCUSSION TITLE
RandSpatialCropSamples seems to output a strange result

### DISCUSSION THREAD

**Discussion by patriciacs1994** [Project-MONAI/MONAI#2052] [Q&A]
Hello,

I have a 3D image and I want to select a number of random patches with a shape (96,96,96). So I decided to use RandSpatialCropSamples. Here you can see my piece of code:

```
patch_size = 96
train_imtrans = Compose(
    [
        LoadImage(image_only=True),
        ScaleIntensity(),
        AddChannel(),
        RandSpatialCropSamples((96, 96, 96),10, random_size=False), 
        ToTensor(),
    ]
)
train_segtrans = Compose(
    [
        LoadImage(image_only=True),
        AddChannel(),
        RandSpatialCropSamples((96, 96, 96),10, random_size=False), 
        ToTensor(),
    ]
)
val_imtrans = Compose([LoadImage(image_only=True), RandSpatialCropSamples((96, 96, 96),10, random_size=False), ScaleIntensity(), ToTensor()])
val_segtrans = Compose([LoadImage(image_only=True),RandSpatialCropSamples((96, 96, 96),10, random_size=False), ToTensor()])


# create a training data loader
train_ds = ArrayDataset(train_img_list[0], train_imtrans, train_img_list[1], train_segtrans)
train_loader = DataLoader(train_ds, batch_size=2, num_workers=8, pin_memory=torch.cuda.is_available())

# create a validation data loader
val_ds = ArrayDataset(val_img_list[0], val_imtrans, val_img_list[1], val_segtrans)
val_loader = DataLoader(val_ds, batch_size=1, num_workers=4, pin_memory=torch.cuda.is_available())

for batch_data in train_loader:
    inputs, labels = batch_data[0], batch_data[1]
```

My problem is in the **inputs** and **labels** parameters. both have a size of (2,1,96,96,96), but when I plot them, both inputs and labels are pieces of the original image and labels should be the correspondent piece of GT image but it is not. I realized that batch_data has an output size of (num_samples*2). Am I doing something wrong or is there an error in the function?

NOTE: train_image_list and val_img_list are lists of lists of strings. with size (2,56)

Thank you in advance for your help

**Comment by Nic-Ma** [comment]
Hi @patriciacs1994 ,

May I know what `DataLoader` you are using?
If you use the `torch.utils.data.DataLoader`, please try with `monai.data.DataLoader`, we have enhanced `collate_fn` callback.

Thanks.

**Reply by patriciacs1994** [reply]
Hi @Nic-Ma 
I was using torch.utils.data.DataLoader.
How I can use collate_fn?

Thank you in advance.

**Comment by rijobro** [comment]
Hi @patriciacs1994, I think that by using `ArrayDataset`, you might be making life slightly difficult for yourself. I would advise switching to using our normal `Dataset` (or `CacheDataset` etc), and using a dictionary of data, with dictionary transforms. We do this in most of our segmentation tutorials, for example [this brats one](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/brats_segmentation_3d.ipynb).

The problem you're seeing here is sneaky. If you executed `len(batch_data)`, I think it would return 20. This corresponds to two input types (image and label) and 10 samples produced by `RandSpatialCropSamples`. Each of the elements would then have batch size of 2, making things even messier. I think if you were to take `batch_data[i]` and `batch_data[i+10]`, they would correspond with each other. This is why @Nic-Ma suggested playing with the collation function. However, I think it would be simpler in your case to just switch to using a dictionary of data.

**Reply by rijobro** [reply]
Just opened a related issue: https://github.com/Project-MONAI/MONAI/issues/2111.

**Reply by patriciacs1994** [reply]
Thank you for your answers. 
Changing the format of the dataset to a dictionary solves all the problems.
